src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs

```diff
@@ -26,72 +27,92 @@ public static unsafe void CreateDirectory(string fullPath, byte[]? securityDescr
                 return;
             }
 
-            List<string> stackDir = new List<string>();
+            fixed (byte* pSecurityDescriptor = securityDescriptor)
+            {
+                Interop.Kernel32.SECURITY_ATTRIBUTES secAttrs = new Interop.Kernel32.SECURITY_ATTRIBUTES
+                {
+                    nLength = (uint)sizeof(Interop.Kernel32.SECURITY_ATTRIBUTES),
+                    lpSecurityDescriptor = (IntPtr)pSecurityDescriptor
+                };
 
-            // Attempt to figure out which directories don't exist, and only
-            // create the ones we need.  Note that FileExists may fail due
-            // to Win32 ACL's preventing us from seeing a directory, and this
-            // isn't threadsafe.
+                // Every call to CreateDirectory uses EnsureExtendedPrefix on the supplied path, which
+                // creates a new string if the prefix was missing.  Since we're making at least one and
+                // potentially multiple calls to CreateDirectory, we lift out that prefixing so that the
+                // string that's created can be shared across all calls rather than having a temporary in each.
+                fullPath = PathInternal.EnsureExtendedPrefix(fullPath);
+
+                // We know the directory doesn't exist.  Before doing more work to parse the path into
+                // subdirectories and checking each one for existence, try to just create the directory.
+                // If it works, which is a common case when the parent directory already exists, we're done.
+                // Otherwise, ignore the error and proceed with the full handling. This makes a fast path
+                // faster and a slow path a little bit slower.
+                if (Interop.Kernel32.CreateDirectory(fullPath, &secAttrs))
+                {
+                    return;
+                }
 
-            bool somepathexists = false;
-            int length = fullPath.Length;
+                // Use enough stack space to hold three directory paths.  After that, ValueListBuilder
+                // will grow with ArrayPool arrays.
+                ThreeObjects scratch = default;
+                var stackDir = new ValueListBuilder<string>(MemoryMarshal.CreateSpan(ref Unsafe.As<object, string>(ref scratch.Arg0!), 3));
 
-            // We need to trim the trailing slash or the code will try to create 2 directories of the same name.
-            if (length >= 2 && PathInternal.EndsInDirectorySeparator(fullPath.AsSpan()))
-            {
-                length--;
-            }
+                // Attempt to figure out which directories don't exist, and only
+                // create the ones we need.  Note that FileExists may fail due
+                // to Win32 ACL's preventing us from seeing a directory, and this
+                // isn't threadsafe.
 
-            int lengthRoot = PathInternal.GetRootLength(fullPath.AsSpan());
+                bool somepathexists = false;
```

Comment:
nit: I know it was named like this before your changes

```suggestion
                bool somePathExists = false;
```